torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,171 @@
-    }
+      // Decay the first and second moment running average coefficient
+      exp_avg.mul_(options.beta1()).add_(grad, 1 - options.beta1());
+      exp_avg_sq.mul_(options.beta2()).addcmul_(grad, grad, 1 - options.beta2());
 
-    const auto step_size =
-        options.learning_rate() / bias_correction1;
+      Tensor denom;
+      if(options.amsgrad()) {
+        // Maintains the maximum of all 2nd moment running avg. till now
+        at::max_out(max_exp_avg_sq, exp_avg_sq, max_exp_avg_sq);
+        // Use the max. for normalizing running avg. of gradient
+        denom = (max_exp_avg_sq.sqrt() / sqrt(bias_correction2)).add_(options.eps());
+      } else {
+        denom = (exp_avg_sq.sqrt() / sqrt(bias_correction2)).add_(options.eps());
+      }
 
-    NoGradGuard guard;
-    p.addcdiv_(exp_average, denom.sqrt() + options.eps(), -step_size);
+      auto step_size = options.lr() / bias_correction1;
+      p.data().addcdiv_(exp_avg, denom, -step_size);
+    }
   }
 }
 
+void Adam::add_parameters(const std::vector<Tensor>& parameters) {
+  param_groups_.emplace_back(OptimizerParamGroup(parameters, defaults_->clone()));
+}
+
+const std::vector<Tensor>& Adam::parameters() const noexcept {
+  return param_groups_.at(0).params();
+}
+
+std::vector<Tensor>& Adam::parameters() noexcept {
+  return param_groups_.at(0).params();
+}
+
+size_t Adam::size() const noexcept {
+  return _size_new_design();
+}
+
 void Adam::save(serialize::OutputArchive& archive) const {
   serialize(*this, archive);
 }
 
 void Adam::load(serialize::InputArchive& archive) {
-  serialize(*this, archive);
+  IValue pytorch_version;
+  if (archive.try_read("pytorch_version", pytorch_version)) {
+    serialize(*this, archive);
+  }
+  else { // deserializing archives saved in old format (prior to version 1.5.0)
+    TORCH_WARN(
+      "Your serialized Adam optimizer is still using the old serialization format. "
+      "You should re-save your Adam optimizer to use the new serialization format.");
+    std::vector<int64_t> step_buffers;
+    std::vector<at::Tensor> exp_average_buffers;
+    std::vector<at::Tensor> exp_average_sq_buffers;
+    std::vector<at::Tensor> max_exp_average_sq_buffers;
+    torch::optim::serialize(archive, "step_buffers", step_buffers);
+    torch::optim::serialize(archive, "exp_average_buffers", exp_average_buffers);
+    torch::optim::serialize(archive, "exp_average_sq_buffers", exp_average_sq_buffers);
+    torch::optim::serialize(archive, "max_exp_average_sq_buffers", max_exp_average_sq_buffers);
+    // since there were no param_groups prior to version 1.5.0, assuming all tensors are now in one param_group
+    std::vector<Tensor> params = param_groups_.at(0).params();
+    for (size_t idx = 0; idx < step_buffers.size(); idx++) {
+      auto state = std::make_unique<AdamParamState>();
+      state->step(step_buffers[idx]);
+      state->exp_avg(exp_average_buffers[idx]);
+      state->exp_avg_sq(exp_average_sq_buffers[idx]);
```

Comment:
To index a vector in this function, I'd recommend using `step_buffers.at(idx)` instead of `step_buffers[idx]`. The reason is that `step_buffers.at(idx)` checks whether `idx` is out of bound for `step_buffers`, and gives a meaningful error message when the out-of-bound error happens (versus `step_buffers[idx]` which returns garbage data if `idx` is out of bound)